portal-impl/src/com/liferay/portal/service/impl/ClassNameLocalServiceImpl.java

```diff
@@ -113,15 +113,23 @@ public ClassName getClassName(String value) {
 
 		ClassName className = _classNames.get(value);
 
-		if (className == null) {
-			try {
-				className = classNameLocalService.addClassName(value);
-			}
-			catch (Throwable t) {
-				className = classNameLocalService.fetchClassName(value);
+		if (className != null) {
+			return className;
+		}
+
+		synchronized (value.intern()) {
```

Comment:
@shuyangzhou 
The put into local cache can only be done when the data is inserted into database successfully, this is to protect that we do not invoke adding to database more than one time, so that to avoid the index conflicts.  

Tina